### PR TITLE
DB-10410: make data dictionary backward compatible

### PR DESCRIPTION
**Pull request by jyuanca**
Please also review ee branch

### CODE REVIEW

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: splice_machine/src/test/java/com/splicemachine/homeless/SerializationTest.java [hunk 1]
```diff
@@ -20,12 +20,14 @@
...
 import org.apache.spark.sql.types.StructField;
 import org.junit.Assert;
+import org.junit.Ignore;
 import org.junit.Test;
 
 import java.io.*;
 import java.sql.Types;
 import java.util.Arrays;
 
+@Ignore
```
Comment (martinrupp) [inline splice_machine/src/test/java/com/splicemachine/homeless/SerializationTest.java:30]: there's some tests here the Assert that written objects have a specific binary representation, these changed i guess? but there's also changes that just assume we can still deserialize old representations (e.g. testTriggerInfo), i think that should still be the case, or we wouldn't be able to upgrade, or am i wrong? So i would suggest to not @Ignore these tests, but instead adjust them. I can help

**Review by martinrupp — CHANGES_REQUESTED** [review]

File: splice_machine/src/test/java/com/splicemachine/homeless/SerializationTest.java (hunk 1 above)
Comment (jyuanca) [reply splice_machine/src/test/java/com/splicemachine/homeless/SerializationTest.java:30]: Good catch! I temporarily disabled some of the tests and forgot to enable or modify. Will look into this. Thanks!

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2230/

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2233/

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2234/

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2239/

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2240/

**Comment by martinrupp** [comment]
i justed tested this with the check_upgrade.sh script from DB-11134 (upgrading from 3.1.0.1971) and it works, running 8 upgrade scripts, including UpgradeStoredObjects. The DB is usable afterwards. Will read the code now

File: db-engine/src/main/java/com/splicemachine/db/iapi/types/SQLChar.java [hunk 2]
```diff
@@ -1179,8 +1124,169 @@ private void resetForMaterialization() {
...
+        // user data and finally write an EOF-marker is required.
+        header.generateInto(out, utflen);
+        writeUTF(out, strlen, isRaw, null );
+        header.writeEOF(out, utflen);
+    }
+
+    @Override
+    protected void readExternalNew(ObjectInput in) throws IOException {
+        byte[] bs = ArrayUtil.readByteArray(in);
+        ExtensionRegistry extensionRegistry = ExtensionRegistry.newInstance();
```
Comment (martinrupp) [inline db-engine/src/main/java/com/splicemachine/db/iapi/types/SQLChar.java:1244]: is the ExtensionRegistry re-created for every readXNew? is that efficient? shouldn't we maybe have a ExtensionRegistry extensionRegistry as a parameter to readExternalNew, and have these things intialized once?

**Review by martinrupp — COMMENTED** [review]
this looks great! only concern i would have is about performance (see comment). have you made any checks how this behaves when starting from a larger database?

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2272/

**Comment by cloudspliceci** [comment]
TEST FAILED
https://jenkins.build.splicemachine-dev.io/job/AutomatedTesting/job/spliceengine-PR/115/

File: db-engine/src/main/java/com/splicemachine/db/iapi/types/SQLChar.java (hunk 2 above)
Comment (jyuanca) [reply db-engine/src/main/java/com/splicemachine/db/iapi/types/SQLChar.java:1244]: Good point. Fixed

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2326/

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2328/

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: db-engine/src/main/java/com/splicemachine/db/iapi/types/SQLChar.java (hunk 2 above)
Comment (martinrupp) [reply db-engine/src/main/java/com/splicemachine/db/iapi/types/SQLChar.java:1244]: i don't see this being resolved, am i missing something?

**Review by martinrupp — CHANGES_REQUESTED** [review]
i made a comment about the ExtensionRegistry being re-created every time, and you commented `fixed` but it looks like it's still the same. am i missing something?

have you made any checks for performance when starting from a larger database?

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2376/

**Comment by martinrupp** [comment]
reusing ExtensionRegistry looks good. I would really suggest to check if we have performance penalties

**Review by martinrupp — CHANGES_REQUESTED** [review]
reusing ExtensionRegistry looks good. I would really suggest to check if we have performance penalties

File: db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/ColumnDescriptor.java [hunk 3]
```diff
@@ -268,7 +268,8 @@ public ColumnDescriptor(String columnName,
         this.uuid = uuid;
         this.defaultUUID = defaultUUID;
         this.collectStatistics = collectStats;
-        assertAutoinc(autoincInc!=0, autoincInc, columnDefaultInfo);
+        // TODO : uncomment
```
Comment (ipraznik-splice) [inline db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/ColumnDescriptor.java:271]: should it be uncommented? If not please remove the comment

**Review by ipraznik-splice — CHANGES_REQUESTED** [review]
please check //TODO uncomment

**Comment by jyuanca** [comment]
TPCC benchmark results http://nexus.splicemachine.com:8080/view/test_builds/job/test_onDemand/view/benchmarks/job/tpcc_no_version_switch/56/artifact/artifacts/index.html

Compared to weekly TPCC, no significant differences
http://nexus.splicemachine.com:8080/view/test_builds/job/test_onDemand/view/benchmarks/job/perf_weekly_tpcc/lastSuccessfulBuild/artifact/artifacts/index.html

**Comment by jyuanca** [comment]
@jpanko1 Please rebase to latest to pick up changes from DB-10044

**Comment by martinrupp** [comment]
> TPCC benchmark results http://nexus.splicemachine.com:8080/view/test_builds/job/test_onDemand/view/benchmarks/job/tpcc_no_version_switch/56/artifact/artifacts/index.html
> 
> Compared to weekly TPCC benchmark results, no significant differences
> http://nexus.splicemachine.com:8080/view/test_builds/job/test_onDemand/view/benchmarks/job/perf_weekly_tpcc/lastSuccessfulBuild/artifact/artifacts/index.html

this is great, but would TPCC even go through that code? i thought your code would only be called when shutting down the database and restart again? Like, is it possible the restart would take longer?

**Review by ipraznik-splice — APPROVED** [review]

**Comment by jyuanca** [comment]
@martinrupp TPCC has a lot of small transactions that access data dictionary. Data dictionary does not have a significant impact on startup/shutdown performance.

**Review by martinrupp — APPROVED** [review]

**Review by carolp-503 — APPROVED** [review]
+1

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Review by martinrupp — CHANGES_REQUESTED** [review]
please hold, i think i found upgrade problems (more details follow)

**Comment by martinrupp** [comment]
i fail to do an upgrade for this one:

```
git checkout origin/master
./start-splice-cluster
./sqlshell.sh
```

execute this SQL
```
create table a ( i int);
create table b ( i int);
create trigger tg2 after insert on a  REFERENCING NEW AS NEW_ROW for each row when (NEW_ROW.i = 1) begin atomic insert into b values 1; insert into b values 2; end;
create trigger tg3 after insert on b  REFERENCING NEW AS NEW_ROW for each row when (NEW_ROW.i = 3) begin atomic insert into a values 1; insert into b values 2; end;
quit;
```

cherry-pick DB-10410 and restart
```
./start-splice-cluster -k
git cherry-pick origin/DB-10410
./start-splice-cluster -l
```

fails to start region server. `platform_it/spliceRegionSvr2.log
[spliceRegionSvr2.log](https://github.com/splicemachine/spliceengine/files/6042536/spliceRegionSvr2.log)
`, see attached log

**Comment by jyuanca** [comment]
It should work. Try this to run upgrade test for standalone. 
```
git checkout origin/master
./start-splice-cluster
./sqlshell.sh
```

execute this SQL
```
create table a ( i int);
create table b ( i int);
create trigger tg2 after insert on a  REFERENCING NEW AS NEW_ROW for each row when (NEW_ROW.i = 1) begin atomic insert into b values 1; insert into b values 2; end;
create trigger tg3 after insert on b  REFERENCING NEW AS NEW_ROW for each row when (NEW_ROW.i = 3) begin atomic insert into a values 1; insert into b values 2; end;
quit;
```
move platform_it/target/hbase and platform_it/target/zookeeper to a temp directory

cherry-pick DB-10410 and rebuild
mvn clean install -Pcore,cdh6.3.0,ee -DskipTests
move platform_it/target/hbase and platform_it/target/zookeeper back
restart splice
```
./start-splice-cluster -b
```

**Comment by martinrupp** [comment]
hi jun, i checked again and you're right. i really have to clean build in between. will now use attached script for checking upgrades manually
[upgradetest.txt](https://github.com/splicemachine/spliceengine/files/6044424/upgradetest.txt)

**Review by martinrupp — APPROVED** [review]

**Comment by jyuanca** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**